engine/pkg/rpcerror/normalize.go

```diff
@@ -0,0 +1,215 @@
+	var opts normalizeOpts
+	for _, op := range ops {
+		op(&opts)
+	}
+
+	if opts.name == "" {
+		var e E
+		opts.name = reflect.TypeOf(e).Name()
+	}
+
+	ret := &Prototype[E]{
+		msg:  opts.msg,
+		name: opts.name,
+	}
+
+	_, exists := prototypeRegistry.LoadOrStore(opts.name, ret)
+	if exists {
+		log.L().Warn("duplicate error type", zap.String("name", opts.name))
+	}
+
+	return ret
+}
+
+// Gen generates a stackless instance of the given error.
+func (p *Prototype[E]) Gen(e *E) error {
+	return &normalizedError[E]{
+		prototype: p,
+		inner:     *e,
+	}
+}
+
+// GenWithStack is similar to Gen but attaches the call stack to it.
+func (p *Prototype[E]) GenWithStack(e *E) error {
+	return errors.WithStack(p.Gen(e))
+}
+
+// Is returns whether errIn is an instance of the prototype.
+func (p *Prototype[E]) Is(errIn error) bool {
+	normalized, ok := tryUnwrapNormalizedError(errIn)
+	if !ok {
+		return false
+	}
+
+	_, ok = normalized.(*normalizedError[E])
+	return ok
+}
+
+// Convert converts errIn to the type argument provided to Normalize.
+func (p *Prototype[E]) Convert(errIn error) (*E, bool) {
+	normalized, ok := tryUnwrapNormalizedError(errIn)
+	if !ok {
+		return nil, false
+	}
+
+	typedNormalized, ok := normalized.(*normalizedError[E])
+	if !ok {
+		return nil, false
+	}
+
+	return &typedNormalized.inner, true
+}
+
+func (p *Prototype[E]) fromJSONBytes(jsonBytes []byte) (typeErasedNormalizedError, error) {
+	var e E
+	if err := json.Unmarshal(jsonBytes, &e); err != nil {
+		return nil, errors.Annotate(err, "failed to unmarshal error info")
+	}
+
+	return &normalizedError[E]{
+		prototype: p,
+		inner:     e,
+	}, nil
+}
+
+type normalizedError[E errorInfo] struct {
+	prototype *Prototype[E]
+	inner     E
+}
+
+func (e *normalizedError[E]) isRetryable() bool {
+	if r, ok := any(e.inner).(retryablity); ok {
+		return r.isRetryable()
+	}
+
+	// Not retryable by default.
+	return false
+}
+
+func (e *normalizedError[E]) name() string {
+	return e.prototype.name
+}
+
+func (e *normalizedError[E]) message() string {
+	return e.prototype.msg
+}
+
+func (e *normalizedError[E]) Error() string {
+	jsonBytes := e.mustMarshalJSON()
+
+	var builder strings.Builder
+	builder.WriteString(e.name())
+	builder.WriteByte(':')
+	builder.WriteByte(' ')
+
+	if e.message() != "" {
+		builder.WriteString(e.message())
+		builder.WriteByte(' ')
+	}
+	builder.Write(jsonBytes)
+	return builder.String()
+}
+
+func (e *normalizedError[E]) mustMarshalJSON() []byte {
+	jsonBytes, err := json.Marshal(e.inner)
+	if err != nil {
+		panic(fmt.Sprintf("marshaling json for %s failed: %s", e.name(), err.Error()))
+	}
+
+	return jsonBytes
+}
+
+func (e *normalizedError[E]) toPB() *pb.ErrorV2 {
+	return &pb.ErrorV2{
+		Name:    e.name(),
+		Details: e.mustMarshalJSON(),
+	}
+}
+
+func (e *normalizedError[E]) statusCode() codes.Code {
+	if statusCoder, ok := any(e.inner).(grpcStatusCoder); ok {
+		return statusCoder.grpcStatusCode()
+	}
```

Comment:
`e.inner` is the generic with `IsErrorInfo` limit. 
Why we need another `grpcStatusCoder ` method here? 
Seems strange.

Reply:
We might not limit the usage of this package to processing gRPC-related error. Multiple interfaces seem good for decoupling.